lib/jsonwp-proxy/proxy.js

```diff
@@ -232,15 +233,108 @@ class JWProxy {
     }
   }
 
+  async getDownstreamProtocol () {
+    const [, resBody] = await this.proxy('/status', 'GET');
+    return this.getProtocolFromResBody(resBody);
+  }
+
+  requestToCommandName (url, method) {
+    const extractCommandName = (pattern) => {
+      const pathMatch = pattern.exec(url);
+      return pathMatch ? routeToCommandName(pathMatch[1], method) : null;
+    };
+    let commandName = routeToCommandName(url, method);
+    if (!commandName && _.includes(url, '/wd/hub/session/')) {
+      commandName = extractCommandName(/\/wd\/hub\/session\/[^\/]+(.+)/);
+    }
+    if (!commandName && _.includes(url, '/wd/hub/')) {
+      commandName = extractCommandName(/\/wd\/hub(\/.+)/);
+    }
+    return commandName;
+  }
+
+  async proxyCommand (url, method, body = null) {
+    const commandName = this.requestToCommandName(url, method);
+    if (!commandName) {
+      return await this.proxy(url, method, body);
+    }
+    log.debug(`Matched '${url}' to command name '${commandName}'`);
+
+    // Handle "crossing" endpoints for the case
+    // when upstream and downstream drivers operate different protocols
+
+    // Same url, but different arguments
+
+    if (commandName === 'timeouts')  {
+      return await this.proxySetTimeouts(url, method, body);
+    }
+
+    // Same arguments, but different URLs
+
+    const conflicts = [
+      {
+        commandNames: ['execute', 'executeAsync'],
+        jsonwpConverter: () => url.replace(/\/execute.*/,
+          url.includes('async') ? '/execute_async' : '/execute'),
+        w3cConverter: () => url.replace(/\/execute.*/,
+          url.includes('async') ? '/execute/async' : '/execute/sync'),
+      },
+      {
+        commandNames: ['getElementScreenshot'],
+        jsonwpConverter: () => url.replace(/\/element\/([^\/]+)\/screenshot$/,
+          '/screenshot/$1'),
+        w3cConverter: () => url.replace(/\/screenshot\/(.+)$/,
+          '/element/$1/screenshot'),
+      },
+    ];
+
+    for (const {commandNames, jsonwpConverter, w3cConverter} of conflicts) {
+      if (!commandNames.includes(commandName)) {
+        continue;
+      }
+
+      try {
+        log.info(`The '${url}' API might not be the same for the downstream protocol. ` +
+          'Trying to proxy as is');
+        return await this.proxy(url, method, body);
+      } catch (e) {
+        let downstreamProtocol;
```

Comment:
Are there cases where the downstream protocol changes for a given proxy?

Reply:
Practically not, but theoretically it is possible. Which approach do you like more?

Reply:
If it doesn't happen, either the protocol should be saved the first time this code is run, or the proxy should be retrieved through a factory function (so it can be async) which does the single call to figure out the protocol (or, if we know the protocol beforehand, takes the protocol and doesn't do the check), and saves it, so every call can be rewritten correctly.

Reply:
ok, I'm caching it now